src/mist/io/methods.py

```diff
@@ -1483,26 +1483,40 @@ def get_machine_actions(machine_from_api, conn, extra):
 
 def list_machines(user, cloud_id):
     """List all machines in this cloud via API call to the provider."""
-    cloud = Cloud.objects.get(owner=user, id=cloud_id)
+    try:
+        cloud = Cloud.objects.get(owner=user, id=cloud_id)
+    except Cloud.DoesNotExist:
+        raise NotFoundError("Unknown cloud with id %s" % cloud_id)
     try:
         conn = connect_provider(cloud)
-        machines = conn.list_nodes()
+        machines_from_provider = conn.list_nodes()
     except InvalidCredsError:
         raise CloudUnauthorizedError()
     except Exception as exc:
         log.error("Error while running list_nodes: %r", exc)
         raise CloudUnavailableError(exc=exc)
+
+    # create a dict with the machines from the db with key the machine_id
+    machines_from_db = {machine.machine_id: machine
+                        for machine in Machine.objects(cloud=cloud)}
+    now = datetime.utcnow()
     ret = []
-    for m in machines:
+
+    for m in machines_from_provider:
+
+        machine_entry = machines_from_db.pop(m.id, None)
+        if machine_entry:
+            machine_entry.update(set__last_seen=now, set__missing_since=None)
```

Comment:
Why not just set the attributes and the call `.save()`? Doing this using `update` and `set__` atomic operations is not really that more efficient (you have already loaded the machine model instance) and is harder to read. Also, atomic operations such as this bypass any type validation/sanitization that the model usually provides, as we found out the bitter way with the notorious bug we had concerning machine id's.

So, I'm not saying that this is wrong, but why not keep it simple?